services/venice-controller/src/test/java/com/linkedin/venice/controller/TestVeniceControllerStateModel.java

```diff
@@ -126,8 +135,287 @@ public void testStateModelClose() {
+    AtomicBoolean taskCompleted = new AtomicBoolean(false);
+
+    VeniceControllerClusterConfig mockClusterConfig = mock(VeniceControllerClusterConfig.class);
+    when(mockClusterConfig.isVeniceClusterLeaderHAAS()).thenReturn(false);
+
+    VeniceControllerStateModel testStateModel = new VeniceControllerStateModel(
+        "test-cluster",
+        mock(ZkClient.class),
+        mock(HelixAdapterSerializer.class),
+        mockMultiClusterConfig,
+        mock(VeniceHelixAdmin.class),
+        mock(MetricsRepository.class),
+        mock(ClusterLeaderInitializationRoutine.class),
+        mock(RealTimeTopicSwitcher.class),
+        Optional.empty(),
+        mock(HelixAdminClient.class),
+        Optional.empty());
+
+    testStateModel.setClusterConfig(mockClusterConfig);
+    // Set a very short timeout (100ms) so the test doesn't take long
+    testStateModel.setStateTransitionTimeout(100);
+
+    // Use spy to override initClusterResources()
+    VeniceControllerStateModel spyStateModel = spy(testStateModel);
+
+    // Mock initHelixManager to set a mock helix manager (avoid real Helix connection)
+    SafeHelixManager mockHelixManager = mock(SafeHelixManager.class);
+    when(mockHelixManager.isConnected()).thenReturn(true);
+    when(mockHelixManager.getInstanceName()).thenReturn("test-instance");
+    doAnswer(invocation -> {
+      spyStateModel.setHelixManager(mockHelixManager);
+      return null;
+    }).when(spyStateModel).initHelixManager("test-controller");
+
+    // Mock initClusterResources to take longer than timeout
+    doAnswer(invocation -> {
+      taskStarted.countDown();
+      try {
+        // Sleep longer than timeout to trigger TimeoutException
+        for (int i = 0; i < 100; i++) {
+          if (Thread.currentThread().isInterrupted()) {
+            taskWasCancelled.set(true);
+            LOGGER.info("Background task was cancelled by timeout catch block");
+            return null;
+          }
+          Thread.sleep(100);
+        }
+        taskCompleted.set(true);
+      } catch (InterruptedException e) {
+        taskWasCancelled.set(true);
+        LOGGER.info("Background task was interrupted by timeout catch block");
+        Thread.currentThread().interrupt();
+      }
+      return null;
+    }).when(spyStateModel).initClusterResources();
+
+    Message message = mock(Message.class);
+    when(message.getTgtName()).thenReturn("test-controller");
+    when(message.getFromState()).thenReturn("STANDBY");
+    when(message.getToState()).thenReturn("LEADER");
+    when(message.getResourceName()).thenReturn("test-cluster_0");
+
+    // This should throw VeniceException due to timeout, and the catch block should cancel the future
+    expectThrows(
+        VeniceException.class,
+        () -> spyStateModel.onBecomeLeaderFromStandby(message, mock(NotificationContext.class)));
+
+    // Wait for task to start
+    assertTrue(taskStarted.await(2, TimeUnit.SECONDS), "Task should have started");
```

Comment:
Could we give a longer wait time to reduce flakiness, 2 seconds feel a bit too short.m same for the following test,`testStateTransitionExecutionExceptionCancelsBackgroundTask`